src/main/java/com/mongodb/kafka/connect/source/MongoCopyDataManager.java

```
@@ -0,0 +1,162 @@
+ * </ol>
+ */
+class MongoCopyDataManager implements AutoCloseable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MongoCopyDataManager.class);
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final AtomicBoolean errored = new AtomicBoolean();
+    private volatile String errorMessage;
+    private final AtomicInteger namespacesToCopy;
+    private final MongoSourceConfig sourceConfig;
+    private final MongoClient mongoClient;
+    private final ExecutorService executor;
+    private final ConcurrentLinkedQueue<BsonDocument> queue = new ConcurrentLinkedQueue<>();
+
+    MongoCopyDataManager(final MongoSourceConfig sourceConfig, final MongoClient mongoClient) {
+        this.sourceConfig = sourceConfig;
+        this.mongoClient = mongoClient;
+
+        String database = sourceConfig.getString(DATABASE_CONFIG);
+        String collection = sourceConfig.getString(COLLECTION_CONFIG);
+
+        List<MongoNamespace> namespaces;
+        if (database.isEmpty()) {
+            namespaces = getCollections(mongoClient);
+        } else if (collection.isEmpty()) {
+            namespaces = getCollections(mongoClient, database);
+        } else {
+            namespaces = singletonList(createNamespace(database, collection));
+        }
+        LOGGER.info("Copying existing data on the following namespaces: {}", namespaces);
+        namespacesToCopy = new AtomicInteger(namespaces.size());
+        executor = Executors.newFixedThreadPool(
+                Math.min(namespacesToCopy.get(), sourceConfig.getInt(COPY_EXISTING_MAX_THREADS_CONFIG)));
+        namespaces.forEach(n -> executor.submit(() -> copyDataFrom(n)));
+    }
+
+    Optional<BsonDocument> poll() {
+        if (errored.get()) {
+            if (!closed.get()) {
+                LOGGER.error(errorMessage);
+                close();
+                throw new ConnectException(errorMessage);
+            }
+            return Optional.empty();
```

I think it would be better to just throw an exception if poll is closed after this has been closed.

---

Done